src/scripts/zoom-controls.js

```diff
@@ -0,0 +1,206 @@
+   */
+  dollyOut(dollyScale) {
+    if (dollyScale === undefined) {
+      dollyScale = this.getZoomScale();
+    }
+
+    if (this.object.isPerspectiveCamera) {
+      this.object.fov = Math.max(this.minFov, Math.min(this.maxFov, this.object.fov / dollyScale));
+      this.object.updateProjectionMatrix();
+    }
+    else if (this.object.isOrthographicCamera) {
+      this.object.zoom = Math.max(this.minZoom, Math.min(this.maxZoom, this.object.zoom / dollyScale));
+      this.object.updateProjectionMatrix();
+    }
+  }
+
+  handleMouseDownDolly(event) {
+    this.dollyStart.set(event.clientX, event.clientY);
+  }
+
+  handleMouseMoveDolly(event) {
+    this.dollyEnd.set(event.clientX, event.clientY);
+
+    this.dollyDelta.subVectors(this.dollyEnd, this.dollyStart);
+
+    if (this.dollyDelta.y > 0) {
+      this.dollyIn();
+    } else if (this.dollyDelta.y < 0) {
+      this.dollyOut();
+    }
+
+    this.dollyStart.copy(this.dollyEnd);
+  }
+
+  handleMouseWheel(event) {
+    if (event.deltaY < 0) {
+      this.dollyIn(this.getZoomScale());
+    } else if (event.deltaY > 0) {
+      this.dollyOut(this.getZoomScale());
+    }
+  }
+
+  handleTouchStartDolly(event) {
+    var dx = event.touches[0].pageX - event.touches[1].pageX;
+    var dy = event.touches[0].pageY - event.touches[1].pageY;
+
+    var distance = Math.sqrt(dx * dx + dy * dy);
+
+    this.dollyStart.set(0, distance);
+  }
+
+  handleTouchMoveDolly(event) {    
+    var dx = event.touches[0].pageX - event.touches[1].pageX;
+    var dy = event.touches[0].pageY - event.touches[1].pageY;
+
+    var distance = Math.sqrt(dx * dx + dy * dy);
+
+    this.dollyEnd.set(0, distance);
+
+    this.dollyDelta.set(0, Math.pow(this.dollyEnd.y / this.dollyStart.y, this.zoomSpeed));  
+      
+    if (this.dollyDelta.y < 1) {
+      this.dollyOut();
+    } else if (this.dollyDelta.y > 1) {
+      this.dollyIn();
+    }
+  }
+
+  handleTouchEnd(event) {
+    // no-op
+  }
+
+  onTouchStart(event) {
+    if (this.enableZoom === false) return;
+
+    // Only zoom if two fingers are used, pointer-controls will handle one finger movement
+    if (event.touches.length === 2) {
+      this.handleTouchStartDolly(event);
+    }
+  }
+
+  onTouchMove(event) {
+    if (this.enableZoom === false) return;
+
+    event.preventDefault();
+    event.stopPropagation();
+
+    // Only zoom if two fingers are used, pointer-controls will handle one finger movement
+    if (event.touches.length === 2) {
+      this.handleTouchMoveDolly(event);
+    }
+  }
+
+  onTouchEnd(event) {
+    if (this.enableZoom === false) return;
+
+    this.handleTouchEnd(event);
+  }
+
+  onMouseWheel(event) {
+    if (this.enableZoom === false) return;
+
+    event.preventDefault();
+    event.stopPropagation();
+
+    this.handleMouseWheel(event);
+  }
+}
+
+/** @constant {number} ZOOM_MIN Minimum zoom value. */
+ZoomControls.ZOOM_MIN = 1;
+
+/** @constant {number} ZOOM_MAX Maximum zoom value. */
+ZoomControls.ZOOM_MAX = 100;
```

Comment:
We could move these up and inside the class.

Reply:
Yes. We might not need them if we only have PerspectiveCamera on these scenes. Currently we only change the **camera.fov** on zoom, but we can change **camera.zoom** instead. So that's why I have kept them so far.